### PR TITLE
[onnx][quantization] Add JIT pass to insert permutes for conv ops

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30799 Add quantized concat conversion
* **#30679 [onnx][quantization] Add JIT pass to insert permutes for conv ops**

Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18790518](https://our.internmc.facebook.com/intern/diff/D18790518)

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 1]
```diff
@@ -72,18 +72,20 @@ double getScaleFromInput(Node* input_node) {
...
-    auto tmp = input_node->inputs()[0]->node();
-    return getScaleFromInput(tmp);
-  } else if (input_name == "aten::relu") {
-    auto tmp = input_node->inputs()[0]->node();
-    return getScaleFromInput(tmp);
-  } else if (input_name == "prim::ListUnpack") {
-    return getScaleFromInput(input_node->inputs()[0]->node());
-  } else if (input_name == "aten::split_with_sizes") {
+  else if (
+      input_name == "quantized::max_pool2d" || input_name == "aten::relu" ||
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp]: Can we create a white list (i.e., unordered_set) and check input_name exists in the white list or not?

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -127,8 +127,22 @@ def _optimize_graph(graph, operator_export_type, _disable_torch_constant_prop=Fa
         torch._C._jit_pass_onnx_preprocess_caffe2(graph)
 
         if operator_export_type == OperatorExportTypes.ONNX_ATEN_FALLBACK:
+            torch.onnx.symbolic_helper._quantized_ops.clear()
+            # Unpack quantized weights for conv and linear ops and insert into graph.
             torch._C._jit_pass_onnx_unpack_quantized_weights(graph, params_dict)
 
+            # Insert permutes before and after each conv op to ensure correct order.
+            torch._C._jit_pass_onnx_quantization_insert_permutes(graph, params_dict)
```
Comment (houseroad) [inline torch/onnx/utils.py]: Is it possible we can use _jit_pass_custom_pattern_based_rewrite_graph to insert permutes as well?

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 3]
```diff
@@ -248,5 +250,55 @@ void UnpackQuantizedWeights(
...
+  script::parseIR(pattern, &pattern_graph, vmap);
+
+  const auto& matches = findPatternMatches(pattern_graph, *graph);
+
+  for (const auto& match : matches) {
+    auto match_vmap = match.values_map;
+    auto op_node = match_vmap.at(vmap.at("r"))->node();
+    auto input_node = match_vmap.at(vmap.at("r"))->node()->inputs()[0]->node();
+
+    Node* permute_node = graph->create(
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp]: Nit: permute_node_before

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 4]
```diff
@@ -248,5 +250,55 @@ void UnpackQuantizedWeights(
...
+        Symbol::fromQualString("quantized::nchw2nhwc"), {input_node->output()});
+    permute_node->insertBefore(op_node);
+    op_node->removeInput(0);
+    op_node->insertInput(0, permute_node->output());
+
+    Node* permute_node_after = graph->create(
+        Symbol::fromQualString("quantized::nhwc2nchw"), {input_node->output()});
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp]: why use input_node->output()? shall we use op_node->outputs()[0] here?

**Review by houseroad — COMMENTED** [review]
Overall the PR is good. Some inline comments to address.

File: torch/onnx/utils.py (hunk 2 above)
Comment (supriyar) [reply torch/onnx/utils.py]: I did try that but got an assert of the form 
```
terminate called after throwing an instance of 'c10::Error'
  what():  n->owningGraph() == this && n->inBlockList() INTERNAL ASSERT FAILED at ../torch/csrc/jit/ir.h:1191
```
I suspect adding new ops to the graph isn't updating all the required dependencies correctly.

**Review by houseroad — APPROVED** [review]
Looks good. Ship it.

**Pull request closed without merging**